### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by aluminiumgeek**

### CODE REVIEW

**Comment by akariv** [comment]
Please note that this pull-request is pushing an image to docker hub... this should not happen on pull request builds.

File: .travis.sh [hunk 1]
```diff
@@ -12,6 +12,11 @@ elif [ "${1}" == "script" ]; then
     ! docker build --build-arg "CKAN_CLOUD_OPERATOR_IMAGE_TAG=${TAG}" --cache-from viderum/ckan-cloud-operator:jnlp-latest -t ckan-cloud-operator-jnlp -f Dockerfile.jenkins-jnlp . && echo Failed to build jnlp image && exit 1
     echo Great Success! && exit 0
 
+elif [ "${1}" == "test" ]; then
+    echo Run tests
+    docker run --env NO_KUBE_CONFIG=1 --rm --entrypoint '/bin/bash' ckan-cloud-operator -c 'source ~/.bashrc && cd /usr/src/ckan-cloud-operator && ckan-cloud-operator test'
```
Comment (adborden) [inline .travis.sh]: Surprised the `source ~/.bashrc` is necessary. Bash should do this by default unless there's something mis-configured in the container, like `~/.profile` exists.

Comment (adborden) [reply .travis.sh]: Yeah, that seems to be the case.
```
$ docker run --rm --entrypoint /bin/bash ckan-cloud-operator -c "ls -a /root"
.
..
.bashrc
.cache
.conda
.profile
```

Looks like `~/.profile` reads from `~/.bashrc` though, so is the `source` really necessary?

File: Dockerfile [hunk 2]
```diff
@@ -5,12 +5,15 @@ RUN /usr/src/ckan-cloud-operator/docker-build.sh
...
 RUN conda env create -f /environment.yaml
 RUN echo conda activate ckan-cloud-operator >> ~/.bashrc &&\
+    echo 'if [ -z "$NO_KUBE_CONFIG" ]; then' >> ~/.bashrc &&\
     echo '[ -e /etc/ckan-cloud/.kube-config ] && export KUBECONFIG=/etc/ckan-cloud/.kube-config' >> ~/.bashrc &&\
     echo '! [ -z "${KUBE_CONTEXT}" ] && kubectl config use-context "${KUBE_CONTEXT}" >/dev/null 2>&1' >> ~/.bashrc &&\
     echo '! [ "$(ckan-cloud-operator config get --key=ckan-cloud-provider-cluster-main-provider-id --raw)" == "aws" ] && ckan-cloud-operator activate-gcloud-auth >/dev/null 2>&1' >> ~/.bashrc &&\
     echo 'ckan-cloud-operator db proxy port-forward --all-daemon "I know the risks" >/dev/null 2>&1' >> ~/.bashrc &&\
-    echo 'while ! pg_isready -h localhost >/dev/null 2>&1; do sleep .1; done' >> ~/.bashrc
+    echo 'while ! pg_isready -h localhost >/dev/null 2>&1; do sleep .1; done' >> ~/.bashrc &&\
+    echo 'fi' >> ~/.bashrc
```
Comment (adborden) [inline Dockerfile]: These echo's seem kind of error prone. I find that it's easy to mess up the shell quoting or the shell redirection. Why not just put these in a separate file, and then append it? e.g.
```
RUN cat bashrc.inc >> ~/.bashrc
```

File: ckan_cloud_operator/routers/traefik/manager.py [hunk 3]
```diff
@@ -23,7 +23,7 @@ def create(router):
     if dns_provider == 'cloudflare':
         assert cloudflare_email and cloudflare_api_key, 'invalid traefik router spec for cloudflare dns provider'
     # cloudflare credentials are stored in a secret, not in the spec
-    del router_spec['cloudflare']
+    router_spec.pop('cloudflare', None)
```
Comment (adborden) [inline ckan_cloud_operator/routers/traefik/manager.py]: A bit strange syntax for delete, maybe include a comment?

**Review by adborden — COMMENTED** [review]
Hey @aluminiumgeek , hope this review is useful. It helps me to read through and ask questions, so I hope that it doesn't impede your work to do so.

Are these tests meant solely to be run by CI? I tried running the tests locally but it just hung. Not sure if this is something to be documented or not.

```
$ docker run --env NO_KUBE_CONFIG=1 --rm --entrypoint '/bin/bash' ckan-cloud-operator -c 'source ~/.bashrc && cd /usr/src/ckan-cloud-operator && ckan-cloud-operator test'
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

File: Dockerfile (hunk 2 above)
Comment (aluminiumgeek) [reply Dockerfile]: I agree

**Comment by aluminiumgeek** [comment]
@adborden 
> Are these tests meant solely to be run by CI? 

No. It works well on localhost. Ensure you have freshly built `ckan-cloud-operator`, or you could run it even from conda env without docker, or you could just run `coverage run -m unittest discover` directly.

Added a note about testing to README.

File: ckan_cloud_operator/routers/traefik/manager.py [hunk 4]
```diff
@@ -23,7 +23,8 @@ def create(router):
     if dns_provider == 'cloudflare':
         assert cloudflare_email and cloudflare_api_key, 'invalid traefik router spec for cloudflare dns provider'
     # cloudflare credentials are stored in a secret, not in the spec
-    router_spec.pop('cloudflare', None)
+    if 'cloudflare' in router_spec:
+        del router_spec['cloudflare']
```
Comment (adborden) [inline ckan_cloud_operator/routers/traefik/manager.py:27]: Ah, that is much more readable, thanks!

**Review by adborden — APPROVED** [review]
Looks great, thank you for making those changes. I'll have to try out the tests again locally.

**Pull request merged**